lib/src/kraken.dart

```diff
@@ -20,34 +19,73 @@ class KrakenPlugin extends Plugin {
+      });
+    } on LNClientException catch (ex, stacktrace) {
+      plugin.log(level: "broken", message: "error received: ${ex.message}");
+      plugin.log(level: "broken", message: stacktrace.toString());
       rethrow;
     }
   }
 
+  Future<String> handleBolt12(Plugin plugin,
+      {required FetchInvoiceRequest offer}) async {
+    var fetchInvoice = await rpc!
+        .call<FetchInvoiceRequest, FetchInvoiceResponse>(
+            method: "fetchinvoice", params: offer);
+    return fetchInvoice.invoice;
+  }
+
   /// This is the Kraken pay RPC method. This method allows for
   /// an alternate method to pay invoices in Core Lightning and
   /// return payment failure analysis reports.
+  ///
+  /// For more info on what command we are wrapping, check the cln pay command
+  // https://lightning.readthedocs.io/lightning-pay.7.html
   Future<Map<String, Object>> krakenPay(
       Plugin plugin, Map<String, Object> request) async {
-    log(level: "info", message: "This is the kraken pay output.");
+    log(
+        level: "info",
+        message: "This is the kraken pay cmd with request: $request");
     try {
+      String invoice = request["invoice"]! as String;
+      // The human-readable prefix for BOLT 12 offers is lno.
+      if (invoice.startsWith("lno")) {
+        // Modify request with the correct bolt 12!
+        // FIXME: we should remove the msatoshi amount from the request? or the pay command
+        // us smart enough to handle it?
+        request["bolt11"] = await handleBolt12(plugin,
+            offer: FetchInvoiceRequest(
+                offer: invoice, msamsatoshi: request["msatoshi"] as String?));
+      } else {
+        request["bolt11"] = invoice;
+      }
       // TODO: rpc should return an exception if any error occurs, and if we have the error returned run the doctor command
-      HashMap<String, Object> result =
-          await rpc!.call(method: "pay", params: PayRequest.fromJson(request));
-      return Future.value(result);
+      PayResponse result = await rpc!.call<PayRequest, PayResponse>(
+          method: "pay",
+          params: PayRequest.fromJson(request),
+          onDecode: (jsonPayload) =>
+              PayResponse.fromJson(jsonDecode(jsonEncode(jsonPayload))));
+      return result.toJSON() as Map<String, Object>;
+    } on LNClientException catch (ex, stacktrace) {
+      request["raw_reason"] = ex.message;
+      plugin.log(level: "broken", message: "error received: ${ex.message}");
+      plugin.log(level: "broken", message: stacktrace.toString());
+      return krakenDoctor(plugin, request);
     } catch (ex, stacktrace) {
       plugin.log(level: "broken", message: "error received: ${ex.toString()}");
-      stderr.write(stacktrace);
+      plugin.log(level: "broken", message: stacktrace.toString());
+      // TODO run the doctor command, and put the CLN exception message inside the doctor request
+      // with identifier `raw_reason`
       rethrow;
     }
```

Comment:
I do not understand this block of code. I can make out that we are trying to log any general exceptions that may occur, but what is the use case? Did you encounter any specific ones?

Reply:
This is a sanity check, I need to remove the TODO comment because it is no longer true.

Debugging the plugin is hard, more log debugging and sanity checks are always better.

Ah! do not need to happen somethings to add a sanity check, but a good manner is to add a sanity check to prevent something happens!